graalpy/graalpy-micronaut-multithreaded/README.md

```diff
@@ -0,0 +1,507 @@
+        size = 0;
+        setPoolSize(2); // ⑤
+    }
+```
+
+❶ The Micronaut [Context](https://docs.micronaut.io/latest/api/io/micronaut/context/annotation/Context) annotation indicates that the `PythonPool` class should be a singleton for the application and is initialized with the application.
+
+❷ A Truffle [Engine](https://www.graalvm.org/truffle/javadoc/org/graalvm/polyglot/Engine.html) is used to share compiled code, which improves warmup and memory footprint.
+
+❸ A [Context](https://www.graalvm.org/truffle/javadoc/org/graalvm/polyglot/Context.html) in Truffle is an isolated GraalPy interpreter state.
+Each thread will pick a context to use for submitted jobs and keep it in a thread local here.
+
+❹ All GraalPy contexts are chained in a simple Deque and the pool just does round-robin scheduling of jobs.
+
+❺ We have randomly picked an initial pool size of 2.
+
+`PythonPool.java`
+```java
+public synchronized void setPoolSize(int newSize) { // ①
+    if (newSize <= 0) {
+        throw new IllegalArgumentException();
+    }
+    while (size > newSize) {
+        try {
+            contexts.takeLast();
+        } catch (InterruptedException e) {
+            throw new RuntimeException(e);
+        }
+        size--;
+    }
+    while (size < newSize) {
+        contexts.addLast(createContext(engine));
+        size++;
+    }
+}
+
+private static Context createContext(Engine engine) {
+    var resourcesDir = Path.of(System.getProperty("user.dir"), "graalpy.resources");
+    if (!resourcesDir.toFile().isDirectory()) { // ②
+        var fs = VirtualFileSystem.create();
+        try {
+            GraalPyResources.extractVirtualFileSystemResources(fs, resourcesDir);
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+    }
+    var context = GraalPyResources.contextBuilder(resourcesDir)
+        .engine(engine)
+        .allowNativeAccess(true) // ③
+        .allowCreateProcess(true) // ④
+        .allowExperimentalOptions(true)
+        .option("python.IsolateNativeModules", "true") // ⑤
+        .build();
+    context.initialize("python");
+    return context;
+}
+```
+
+❶ The `setPoolSize` method can be used to dynamically resize the available GraalPy pool.
+This size directly corresponds to the maximum parallelism of Python code.
+
+❷ The GraalPy plugins for Maven and Gradle ship packages and Python resources in the application Jar.
+However, to use multiple instances of native libraries like NumPy, we need to put them in the real filesystem.
+This could be a temporary directory or any other writable location.
+Here, we are simply extracting to the current working directory if that has not already happened.
+
+❸ GraalPy contexts are locked down by default.
+Since NumPy is a native library, we need to explicitly enable native access.
+
+❹ GraalPy's mechanism for native library isolation relies on being able to spawn helper processes, so we have to enable that capability.
```

Comment:
Maybe ensure the reader that the magic sauce for multi-context is in fact not just simple subprocessing?